### ISSUE TITLE
uber tracking code causes deadlock when changing bad data in DB to good data

### ISSUE THREAD

**Issue by binary1230** [magfest/ubersystem#1881]
This is interesting and bears some further investigation.

backstory: waaaaay waaaaay too late at night the night before opening magstock registration, I was experiencing a reproducible, hard crash of uber that would render the app unresponsive via HTTP.  This is perhaps similar to some other hard crashes we had seen in the past.

@kitsuta and I were hoping that this could be isolated and hardened to not cause uber to become unresponsive.

i had an attendee with attendee.site_number = "asdf" before the migration. when trying to save that attendee with site_number now set to one of the enum ints, the tracking code was trying to resolve the string "asdf" to int, breaking everything.  

This would throw an exception, which would then render the app unresponsive.

before:
```
site_number    = Column(UnicodeText, default='')
```

after:
```
site_number    = Column(Choice(c.CAMPSITE_OPTS), nullable=True, admin_only=True)
```

other suggestion: we should probably in the future run some kind of automated hourly check that just runs the validation checks on every single attendee in the database looking for invalid data in db columns

**Comment by binary1230** [comment]
one more weird component.  I -think- I didn't locally run the DB migration, so my code would have been the correct column type, but the DB would have been the old column type.

**Comment by binary1230** [comment]
ok, tracing this out more, here's the process so far, this can probably be simplified:

For this to work you'll need to either use psql on the command line, or a tool like DataGrip, to modify the DB outside of ubersystem

1) spin up a magstock instance
2) set AT_CON_MODE to True
3) reset uber DB
4) edit attendee named Test Developer, set paid to "doesn't need to"
5) change column type of attendee.site_number from INTEGER to VARCHAR like this:
```sql
ALTER TABLE public.attendee ALTER COLUMN site_number TYPE VARCHAR USING site_number::VARCHAR;
ALTER TABLE public.attendee ALTER COLUMN site_number SET DEFAULT '';
ALTER TABLE public.attendee ALTER COLUMN site_number SET NOT NULL;
```
6) insert bogus data into database using external tool:
```sql
UPDATE Attendee SET site_number='Bogus' WHERE first_name = 'Test';
```

7) Attempt to check in the attendee via /uber/registration

EXPECTED: some errors because we were screwing around with the database, but the app remains responsive via HTTP

ACTUAL: app becomes unresponive and needs a restart


the error that comes up if you're looking at it in pycharm is:

```
2016-06-21 15:49:41,849 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 
2016-06-21 15:59:32,678 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.542575 seconds
2016-06-21 15:59:33,257 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.08947 seconds
2016-06-21 15:59:40,509 [WARNING] uber.models: failed releasing c.BADGE_LOCK on db error; these errors should not happen in the first place and we do not expect releasing the lock to fail when they do, but we still want to keep going if/when this does occur
2016-06-21 15:59:40,603 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.440755 seconds
2016-06-21 15:59:40,891 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.099038 seconds
2016-06-21 16:00:28,254 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.455081 seconds
2016-06-21 16:00:28,571 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.087099 seconds
2016-06-21 16:00:32,349 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.532161 seconds
2016-06-21 16:00:32,645 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.094214 seconds
2016-06-21 16:00:34,889 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.525847 seconds
2016-06-21 16:00:35,206 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.08836 seconds
2016-06-21 16:00:41,444 [WARNING] uber.models: failed releasing c.BADGE_LOCK on db error; these errors should not happen in the first place and we do not expect releasing the lock to fail when they do, but we still want to keep going if/when this does occur
2016-06-21 16:00:41,539 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.465973 seconds
2016-06-21 16:00:41,874 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.090487 seconds
2016-06-21 16:01:07,693 [WARNING] uber.models: failed releasing c.BADGE_LOCK on db error; these errors should not happen in the first place and we do not expect releasing the lock to fail when they do, but we still want to keep going if/when this does occur
2016-06-21 16:01:07,776 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.421715 seconds
2016-06-21 16:01:08,414 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.108411 seconds
2016-06-21 16:01:11,869 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.398597 seconds
2016-06-21 16:01:12,291 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.127484 seconds
2016-06-21 16:02:39,242 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.413838 seconds
2016-06-21 16:02:39,600 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.072175 seconds
2016-06-21 16:03:13,377 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.209756 seconds
2016-06-21 16:03:13,997 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.57815 seconds
2016-06-21 16:03:14,753 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.090173 seconds
2016-06-21 16:03:33,004 [DEBUG] uber.decorators: uber.site_sections.accounts.index loaded in 0.2762 seconds
2016-06-21 16:03:33,967 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.094857 seconds
2016-06-21 16:03:35,998 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.46185 seconds
2016-06-21 16:03:36,872 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.092387 seconds
2016-06-21 16:04:00,201 [DEBUG] uber.decorators: uber.site_sections.registration.check_in_form loaded in 0.139266 seconds
2016-06-21 16:04:14,771 [DEBUG] uber.decorators: uber.site_sections.registration.check_in loaded in 0.077062 seconds
2016-06-21 16:04:14,772 [INFO] cherrypy.error.140256691858456: [21/Jun/2016:16:04:14] HTTP 
Request Headers:
  USER-AGENT: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36
  Remote-Addr: 127.0.0.1
  X-REQUESTED-WITH: XMLHttpRequest
  REFERER: https://localhost:4443/uber/registration/?invalid=True
  Content-Type: application/x-www-form-urlencoded; charset=UTF-8
  CONNECTION: close
  ACCEPT-ENCODING: gzip, deflate, br
  X-FORWARDED-FOR: 10.0.2.2
  COOKIE: _ga=GA1.1.950237344.1453473795; session_id=630bbf8343674db1d291f88aa56d5dedcbec312d
  HOST: localhost
  Content-Length: 150
  ACCEPT-LANGUAGE: en-US,en;q=0.8
  ORIGIN: https://localhost:4443
  X-REAL-IP: 10.0.2.2
  ACCEPT: application/json, text/javascript, */*; q=0.01
2016-06-21 16:04:14,825 [ERROR] cherrypy.error.140256691858456: [21/Jun/2016:16:04:14] HTTP Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1748, in repr
    return repr(dict(column.type.choices).get(int(value), '<nonstandard>'))
ValueError: invalid literal for int() with base 10: 'BOGUS_JUNK'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/lib/encoding.py", line 217, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/_cpdispatch.py", line 61, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 179, in with_timing
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 170, in with_caching
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 194, in with_session
    retval = func(*args, session=session, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 278, in with_restrictions
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 240, in with_rendering
    result = func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 72, in returns_json
    return json.dumps(func(*args, **kwargs), cls=serializer).encode('utf-8')
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/site_sections/registration.py", line 344, in check_in
    session.commit()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 790, in commit
    self.transaction.commit()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 392, in commit
    self._prepare_impl()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 372, in _prepare_impl
    self.session.flush()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2004, in flush
    self._flush(objects)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2031, in _flush
    self.dispatch.before_flush(self, flush_context, objects)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/event/attr.py", line 218, in __call__
    fn(*args, **kw)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1888, in _track_changes
    Tracking.track(action, instance)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1771, in track
    diff = cls.differences(instance)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1761, in differences
    diff[attr] = "'{} -> {}'".format(cls.repr(column, old_val), cls.repr(column, new_val))
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1752, in repr
    raise ValueError('error formatting {} ({!r})'.format(column.name, value)) from e
ValueError: error formatting site_number ('BOGUS_JUNK')

2016-06-21 16:05:10,876 [DEBUG] uber.decorators: uber.site_sections.preregistration.check_prereg loaded in 0.000828 seconds
2016-06-21 16:11:52,341 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.024432 seconds
2016-06-21 16:11:56,671 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.023339 seconds
2016-06-21 16:12:02,333 [DEBUG] uber.decorators: uber.server.index loaded in 0.001097 seconds
2016-06-21 16:12:02,380 [DEBUG] uber.decorators: uber.site_sections.common.index loaded in 0.015459 seconds
2016-06-21 16:12:04,376 [DEBUG] uber.decorators: uber.site_sections.accounts.homepage loaded in 0.087532 seconds
2016-06-21 16:12:05,007 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.08813 seconds
2016-06-21 16:12:07,499 [DEBUG] uber.decorators: uber.site_sections.accounts.logout loaded in 0.001888 seconds
2016-06-21 16:12:07,602 [DEBUG] uber.decorators: uber.site_sections.accounts.login loaded in 0.063666 seconds
2016-06-21 16:12:07,880 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.02847 seconds
2016-06-21 16:12:13,158 [DEBUG] uber.decorators: uber.site_sections.accounts.login loaded in 0.48286 seconds
2016-06-21 16:12:13,347 [DEBUG] uber.decorators: uber.site_sections.accounts.homepage loaded in 0.152517 seconds
2016-06-21 16:12:13,582 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.100256 seconds
2016-06-21 16:12:18,622 [DEBUG] uber.decorators: uber.site_sections.registration.index loaded in 0.484101 seconds
2016-06-21 16:12:19,245 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.159226 seconds
```

notice the scary stuff about c.BADGE_LOCK.

What's going on here is the tracking code is attempting to log the old value ('Bogus') and the new value (an integer).  It causes an exception when trying to cast 'Bogus' to int, which is fine, but this somehow causes the server to stop responding, maybe by not releasing the badge lock?

Will investigate and try to make an easier test case

**Comment by binary1230** [comment]
hang on i think the badge lock stuff is unrelated. verifying.  it seems to be caused when i save the reg form when the attendee contains invalid data, troublehsooting.

**Comment by binary1230** [comment]
ended up getting sidetracked on that badge_lock issue, and you can see all that over here: https://github.com/magfest/ubersystem/pull/1883

so the c.BADGE_LOCK stuff is unrelated (as far as I know)

**Comment by binary1230** [comment]
ok, easier way to look at the crash is for step #7, edit the Test Developer attendee and hit 'save'.

It fails with the following message and requires a restart:

``
2016-06-21 20:07:18,639 [INFO] cherrypy.error: [21/Jun/2016:20:07:18] ENGINE Bus STARTING
2016-06-21 20:07:18,718 [INFO] cherrypy.error.140453913144232: [21/Jun/2016:20:07:18]  Starting WebSocket processing
2016-06-21 20:07:18,722 [INFO] cherrypy.error: [21/Jun/2016:20:07:18] ENGINE Started monitor thread 'Autoreloader'.
2016-06-21 20:07:18,724 [INFO] cherrypy.error: [21/Jun/2016:20:07:18] ENGINE Started monitor thread '_TimeoutMonitor'.
2016-06-21 20:07:18,863 [INFO] cherrypy.error: [21/Jun/2016:20:07:18] ENGINE Serving on http://0.0.0.0:8282
2016-06-21 20:07:19,679 [INFO] cherrypy.error: [21/Jun/2016:20:07:19] ENGINE Bus STARTED
2016-06-21 20:09:52,399 [INFO] cherrypy.error: [21/Jun/2016:20:09:52] ENGINE Started monitor thread 'Session cleanup'.
2016-06-21 20:09:53,526 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 1.122979 seconds
2016-06-21 20:09:54,448 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.020209 seconds
2016-06-21 20:09:58,388 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.521977 seconds
2016-06-21 20:09:58,540 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.020538 seconds
2016-06-21 20:10:17,137 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.487999 seconds
2016-06-21 20:10:17,393 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.026736 seconds
2016-06-21 20:10:36,040 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.557385 seconds
2016-06-21 20:10:36,285 [DEBUG] uber.decorators: uber.server.common_js loaded in 0.02039 seconds
2016-06-21 20:10:42,854 [DEBUG] uber.decorators: uber.site_sections.registration.form loaded in 0.149312 seconds
2016-06-21 20:10:42,962 [ERROR] cherrypy.error.140453913144232: [21/Jun/2016:20:10:42] HTTP Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 194, in with_session
    retval = func(*args, session=session, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 278, in with_restrictions
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 240, in with_rendering
    result = func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 14, in with_check
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/site_sections/registration.py", line 117, in form
    '{} {}'.format(attendee.first_name, attendee.last_name) if c.AT_THE_CON else '')
uber.utils.HTTPRedirect: (['http://localhost/uber/registration/index?uploaded_id=37f91fe4-8964-4f6f-9928-895a66fda90d&message=Test%20Developer%20has%20been%20saved&search_text=Test%20Developer'], 302)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1754, in repr
    return repr(dict(column.type.choices).get(int(value), '<nonstandard>'))
ValueError: invalid literal for int() with base 10: 'Bogus'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/lib/encoding.py", line 217, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/cherrypy/_cpdispatch.py", line 61, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 179, in with_timing
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 170, in with_caching
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 198, in with_session
    session.commit()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 790, in commit
    self.transaction.commit()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 392, in commit
    self._prepare_impl()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 372, in _prepare_impl
    self.session.flush()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2004, in flush
    self._flush(objects)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/orm/session.py", line 2031, in _flush
    self.dispatch.before_flush(self, flush_context, objects)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/sqlalchemy/event/attr.py", line 218, in __call__
    fn(*args, **kw)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1894, in _track_changes
    Tracking.track(action, instance)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1777, in track
    diff = cls.differences(instance)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1767, in differences
    diff[attr] = "'{} -> {}'".format(cls.repr(column, old_val), cls.repr(column, new_val))
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1758, in repr
    raise ValueError('error formatting {} ({!r})'.format(column.name, value)) from e
ValueError: error formatting site_number ('Bogus')

2016-06-21 20:10:42,980 [INFO] cherrypy.error.140453913144232: [21/Jun/2016:20:10:42] HTTP 
Request Headers:
  ORIGIN: https://localhost:4443
  HOST: localhost
  Content-Type: application/x-www-form-urlencoded
  USER-AGENT: Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.103 Safari/537.36
  CACHE-CONTROL: max-age=0
  Content-Length: 645
  ACCEPT-ENCODING: gzip, deflate, br
  REFERER: https://localhost:4443/uber/registration/form
  Remote-Addr: 127.0.0.1
  CONNECTION: close
  COOKIE: _ga=GA1.1.950237344.1453473795; session_id=630bbf8343674db1d291f88aa56d5dedcbec312d
  ACCEPT-LANGUAGE: en-US,en;q=0.8
  X-FORWARDED-FOR: 10.0.2.2
  ACCEPT: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8
  X-REAL-IP: 10.0.2.2
  UPGRADE-INSECURE-REQUESTS: 1

```

**Comment by binary1230** [comment]
OK! getting somewhere.  after we have reached the hung state, if you unset the session_id in the cookie, things still work fine.  if you don't, you get a hang.

Tracing it down into the guts of cherrypy, we're hanging on:
```
cherrypy.serving.session.acquire_lock()
```

So probably in our error handling somewhere, something isn't releasing the lock on the session file as part of the error handling process.  Will investigate more.

**Comment by binary1230** [comment]
Second hunch: 

We're raising an HTTPRedirect exception, and inside that exception we have another exception of type ValueError occurring.

I bet this has something to do with that.

**Comment by binary1230** [comment]
pasted from slack: OK! I think I have diagnosed and fixed that hang we hit onsite at Magstock due to bad DB data.  With those two PR's, we're now ultra-paranoid and crazy in the tracking code.  It turns out that in pages where we're raising an HTTPRedirect, if the tracking code ALSO throws an exception (which it was due to our slightly incorrect DB migration), then cherrypy never releases the lock on the session file.  The next time you try and use the session file (on the next page load), we would hang while trying forever to acquire the lock (which won't happen because nothing ever released it)

**Comment by binary1230** [comment]
final note: for a much easier repro, just throw a fake exception in Tracking.track() on any page that raises a HTTPRedirect()

**Comment by kitsuta** [comment]
Oh wow, this makes so much sense regarding how "random" these deadlocks seemed. We use HTTPRedirect all over the code.

Your current PRs to fix the causes of this bug are good, but is there any way we can also add an exception handler to HTTPRedirect itself, so that we guarantee that we'll never deadlock the server?

**Comment by kitsuta** [comment]
Comment withdrawn in light of https://github.com/magfest/ubersystem/pull/1887, which is exactly the kind of "ultimate failsafe" I was talking about.

**Comment by binary1230** [comment]
ok this is now fixed with #1887
